File: go/difference-of-squares/difference_of_squares.go
// Package diffsquares provides utilities to find the difference between the
// square of the sum and the sum of the squares of the first N natural numbers.
package diffsquares

// SquareOfSum finds the square of the sum of the first n natural numbers.
func SquareOfSum(n int) int {
	sum := Sum(n)
	return sum * sum
}

// SumOfSquares finds the sum of the squares of the first n natural numbers.
func SumOfSquares(n int) int {
	return Sum(n) * (n*2 + 1) / 3
}

// Sum finds the sum of the first n natural numbers.
func Sum(n int) int {
	return n * (n + 1) / 2
}

// Difference finds the difference between the square of the sum of the first
// n natural numbers and the sum of the squares of the first n natural numbers.
func Difference(n int) int {
	return SquareOfSum(n) - SumOfSquares(n)
}

File: solutions/kotlin/difference-of-squares/1/src/main/kotlin/DifferenceOfSquares.kt
data class Squares(val n: Int) {
    private val sumSquares: Int by lazy {
        (1..n).sumOf { it * it }
    }

    private val squareSum: Int by lazy {
        val sum = (1..n).sum()
        sum * sum
    }
    fun sumOfSquares(): Int {
        return sumSquares
    }

    fun squareOfSum(): Int {
        return squareSum
    }

    fun difference(): Int {
        return squareSum - sumSquares
    }
}
